src/WasmParser.ts

```diff
@@ -1780,83 +1785,59 @@ export class BinaryReader {
       this.skipSection();
       return this.read();
     }
-    if (!this.hasVarIntBytes()) {
+    const pos = this._pos;
+    if (!this.hasMoreBytes()) {
       this.state = BinaryReaderState.ELEMENT_SECTION_ENTRY;
       return false;
     }
-    const flags = this.readVarUint7();
-    let tableIndex = 0;
-    if (flags & SegmentFlags.HasTableIndex) {
-      tableIndex = this.readVarUint32();
+    const segmentType = this.readUint8();
+    let mode, tableIndex;
+    switch (segmentType) {
+      case 0x00: // legacy active, funcref externval
+      case 0x04: // legacy active, funcref elemexpr
+        mode = ElementMode.Active;
+        tableIndex = 0;
+        break;
+      case 0x01: // passive, externval
+      case 0x05: // passive, elemexpr
+        mode = ElementMode.Passive;
+        break;
+      case 0x02: // active, externval
+      case 0x06: // active, elemexpr
+        mode = ElementMode.Active;
+        if (!this.hasVarIntBytes()) {
+          this.state = BinaryReaderState.ELEMENT_SECTION_ENTRY;
+          this._pos = pos;
+          return false;
+        }
+        tableIndex = this.readVarUint32();
+        break;
+      case 0x03: // declared, externval
+      case 0x07: // declared, elemexpr
+        mode = ElementMode.Declarative;
+        break;
+      default:
+        throw new Error(`Unsupported element segment type ${segmentType}`);
     }
-
     this.state = BinaryReaderState.BEGIN_ELEMENT_SECTION_ENTRY;
-    this.result = { index: tableIndex };
+    this.result = { mode, tableIndex };
     this._sectionEntriesLeft--;
-    this._segmentFlags = flags;
+    this._segmentType = segmentType;
     return true;
   }
   private readElementEntryBody(): boolean {
-    let funcType = Type.unspecified;
-    const pos = this._pos;
-    if (
-      this._segmentFlags &
-      (SegmentFlags.IsPassive | SegmentFlags.HasTableIndex)
-    ) {
+    let elementType = Type.funcref;
+    if (this._segmentType > 0x00 && this._segmentType < 0x04) {
```

Comment:
Turned this particular one into a `switch` statement, since it's easier to read and reason about this way. However for two other checks I've added helpers.